### PR TITLE
Add version scheme detection task

### PR DESCRIPTION
**Pull request by papiro**
Thanks taking the time to contribute. This template should help guide you through the process of creating a pull request for review. Please erase any part of this template that is not relevant to your pull request:

- [X] Have you followed the guidelines for [contributors](http://docs.libraries.io/contributorshandbook)?
- [X ] Have you checked to ensure there aren't other open pull requests on the repository for a similar change?
- [ ] Is there a corresponding ticket for your pull request?
- [ X] Have you written new tests for your changes?
- [ X] Have you successfully run the project with your changes locally?

In investigating supporting multiple versioning schemes the idea arose to gather some numbers on scheme counts "in the wild".

This PR contains a Rake task.
The purpose of this task is to try to detect (as intelligently as possible) the spread of versioning schemes in use in a list of provided platforms and package names. It is not finished since I am working on consuming the package list from a file right now, but I wanted to submit for review of the algorithm.

### CODE REVIEW

File: lib/tasks/version_scheme_counter.rake [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        global_tallies[:unknown_versions].push([project.platform, project.name, project.versions.pluck(:number)])
+        detected_scheme = :unknown
+      elsif maxes.include?(:semver)
+        detected_scheme = :semver
+      elsif maxes.include?(:calver)
+        detected_scheme = :calver
+      elsif maxes.include?(:maven) && project.platform.downcase == "maven"
+        detected_scheme = :maven
+      elsif maxes.include?(:pep440) && project.platform.downcase == "pypi"
+        detected_scheme = :pep440
```
Comment (papiro) [inline lib/tasks/version_scheme_counter.rake:95]: Need to determine a fall-through tie-breaker

File: lib/tasks/version_scheme_counter.rake [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+#
+# [
+#   ["Maven", "io.redlink.geocoding:proxy-commons"],
+#   ["Packagist", "phraseanet/php-sdk"]
+# ].each { |project| PackageManager.const_get(project[0]).update project[1]}
+#
+#
+# Version scheme detection methodology:
```
Comment (papiro) [inline lib/tasks/version_scheme_counter.rake:9]: The algorithm in plain english

File: lib/tasks/version_scheme_counter.rake [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+          :semver,
+          :pep440,
+          :maven,
+          :calver,
+        ].each do |scheme|
+          if VersionSchemeDetection::VALIDATORS[scheme.upcase].call(version_number)
+            local_tallies[scheme] += 1
+            matchable = true
+          end
+        end
```
Comment (tiegz) [inline lib/tasks/version_scheme_counter.rake:77]: Just to 2x check, this doesn't break on purpose so that we catch all matching schemes for a version? Makes sense, to get the best approximation

File: lib/tasks/version_scheme_counter.rake [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+namespace :version do
+  desc 'Tests a sampling of project\'s versions to count occurrences of different versioning schemes'
+  task scheme_counter: :environment do
+    # projects = Project.order(Arel.sql("RANDOM()")).limit(100)
+
+    global_tallies = tallies.clone.merge({
+      unknown_versions: []
+    })
+
+    Project.find_each(batch_size: 1000) do |project|
```
Comment (tiegz) [inline lib/tasks/version_scheme_counter.rake]: Could do an `includes(:versions)` here to preload

File: lib/tasks/version_scheme_counter.rake [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        end
+      end
+
+      max_count = local_tallies.values.max
+      maxes = local_tallies.select{|_, tally| tally == max_count}.keys
+
+      if maxes.length == 1
+        detected_scheme = maxes[0]
+      elsif maxes.include?(:unknown)
+        global_tallies[:unknown_versions].push([project.platform, project.name, project.versions.pluck(:number)])
```
Comment (tiegz) [inline lib/tasks/version_scheme_counter.rake:90]: If you do the `includes(:versions)` suggestion from above, could change the pluck to a map here too

File: lib/tasks/version_scheme_counter.rake (hunk 1 above)
Comment (tiegz) [reply lib/tasks/version_scheme_counter.rake:99]: Just to 2x here too, the ordering basically determines the priority that we give a certain scheme match, in case there are 2+ schemes that match the most?

Comment (papiro) [reply lib/tasks/version_scheme_counter.rake:99]: yes, that's correct

File: lib/tasks/version_scheme_counter.rake (hunk 3 above)
Comment (papiro) [reply lib/tasks/version_scheme_counter.rake:77]: yep, that's right

File: lib/tasks/version_scheme_counter.rake (hunk 5 above)
Comment (papiro) [reply lib/tasks/version_scheme_counter.rake:90]: What is the difference between the two? I've seen `.map(&:number)` before but I thought it was only needed when referring to a non-getter method on the object?

Comment (tiegz) [reply lib/tasks/version_scheme_counter.rake:90]: The `pluck()` does a new query for the column you need (`SELECT number FROM versions WHERE versions.package_id = ?`) and doesn't hydrate any ActiveRecord models, so it's just a list of values.

So if you don't do the `includes(:versions)` earlier, it won't have any preloaded versions and this'll be the most efficient way.

But if you `includes(:versions)` earlier, it'll preload your versions (and hydrate their ActiveRecord models) with all the numbers in a single query, and it'll already have the `:number` values in memory, so doing a `pluck(:number)` query for each project would be extraneous work.

File: lib/tasks/version_scheme_counter.rake [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      unknown_schemes: []
+    })
+    warnings = []
+
+    packages = CSV.read(args[:package_list])
+    # downcase platform
+    packages.map! { |package| [package[0].downcase, package[1]] }
+
+    # The following makes a query like "(x AND y) OR (x AND y)"
+    package_matcher = "(LOWER(platform) = ? AND name = ?)"
```
Comment (johnbintz-tidelift) [inline lib/tasks/version_scheme_counter.rake:66]: There's a more programmatic and less string-oriented way you can build these SQL statements. We use Arel nodes on the Catalog model in DepCI to build a complex grouping of `OR` queries that is a bit easier to work with: https://github.com/tidelift/dependencyci/blob/main/app/models/catalog.rb#L634-L645

**Review by johnbintz-tidelift — APPROVED** [review]
This looks good! I'll be curious to see what comes out of the analysis.

File: lib/tasks/version_scheme_counter.rake [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    Project
+      .includes(:versions)
+      .where(VersionSchemeDetection.build_project_where_clause(packages))
+      .find_each(batch_size: 1000) do |project|
+      project_platform = project.platform
+      project_name = project.name
+
+      local_tallies = tallies.clone
+
+      if !project.versions_count?
```
Comment (johnbintz-tidelift) [inline lib/tasks/version_scheme_counter.rake:89]: I am surprised that linting tools didn't complain about this one:

```suggestion
      unless project.versions_count?
```
